swupd/cmd/create-pack/main.go

```diff
@@ -0,0 +1,237 @@
+	if fromVersionUint > 0 {
+		fromDir := filepath.Join(stateDir, "www", fromVersion)
+		fromMoM, err := readManifestFile(filepath.Join(fromDir, "Manifest.MoM"))
+		if err != nil {
+			log.Fatalf("couldn't read MoM of FROM_VERSION (%s): %s", fromVersion, err)
+		}
+
+		for _, f := range fromMoM.Files {
+			e, ok := bundles[f.Name]
+			if !ok {
+				// Bundle doesn't exist in new version, no pack needed.
+				continue
+			}
+			if e.to == f.Version {
+				// Versions match, so no pack required.
+				delete(bundles, f.Name)
+				continue
+			}
+			if e.to < f.Version {
+				log.Fatalf("invalid bundle versions for bundle %s, check the MoMs", f.Name)
+			}
+			e.from = f.Version
+		}
+	}
+
+	// TODO: Use goroutines.
+	for _, b := range bundles {
+		pack(stateDir, chrootDir, *outputDir, b, *force)
+	}
+}
+
+type bundleToPack struct {
+	name string
+	from uint32
+	to   uint32
+}
+
+func pack(stateDir, chrootDir string, outputDir string, bundle *bundleToPack, force bool) {
+	toDir := filepath.Join(outputDir, "www", fmt.Sprint(bundle.to))
+	err := os.MkdirAll(toDir, 0755)
+	if err != nil {
+		log.Fatal(err)
+	}
+	outputFilename := filepath.Join(toDir, fmt.Sprintf("pack-%s-from-%d.tar", bundle, bundle.from))
+
+	// If we are not forcing, skip the packs already on disk.
+	if !force {
+		_, err := os.Stat(outputFilename)
+		if err == nil {
+			fmt.Printf("Pack already exists for %s from %d to %d\n", bundle.name, bundle.from, bundle.to)
+			return
+		}
+		if !os.IsNotExist(err) {
+			log.Fatal(err)
+		}
+	}
+
+	fmt.Printf("Packing %s from %d to %d...\n", bundle.name, bundle.from, bundle.to)
+
+	m, err := readManifestFile(filepath.Join(stateDir, "www", fmt.Sprint(bundle.to), "Manifest."+bundle.name))
+	if err != nil {
+		log.Fatal(err)
+	}
+	m.Name = bundle.name
+
+	//
+	// Create the file and write the pack to it.
+	//
+	output, err := os.Create(outputFilename)
+	if err != nil {
+		log.Fatal(err)
+	}
+	info, err := swupd.WritePack(output, m, bundle.from, stateDir, chrootDir)
+	if err != nil {
+		os.RemoveAll(outputFilename)
+		log.Fatal(err)
+	}
+	err = output.Close()
+	if err != nil {
+		os.RemoveAll(outputFilename)
+		log.Fatal(err)
+	}
+
+	//
+	// Output information about the pack produced.
+	//
+	for i, w := range info.Warnings {
+		if i == 0 {
+			fmt.Println("Warnings during pack:")
+		}
+		fmt.Printf("  %s\n", w)
+		if i == len(info.Warnings)-1 {
+			fmt.Println()
+		}
+	}
+	// TODO: Move this to the info structure itself?
+	fullfiles := 0
+	deltas := 0
+	for _, e := range info.Entries {
+		switch e.State {
+		case swupd.PackedFullfile:
+			fullfiles++
+		case swupd.PackedDelta:
+			deltas++
+		}
+	}
+	fmt.Printf("  Fullfiles in pack: %d\n", fullfiles)
+	fmt.Printf("  Deltas in pack: %d\n", deltas)
+}
+
+func parseUint32(s string) uint32 {
+	parsed, err := strconv.ParseUint(flag.Arg(1), 10, 32)
+	if err != nil {
+		log.Fatalf("error parsing value %q: %s", s, err)
+	}
+	return uint32(parsed)
+}
+
+// TODO: Move to manifest.go as an exported function.
```

Comment:
Agree, this should go in manifest.go and used in manifest creation (CreateManifests)

Reply:
This is part of #23.

Reply:
This function can be removed now right?

Reply:
Done. Was waiting #23 getting in.